### ISSUE TITLE
sql: panic while executing 1 statements: SELECT _ FROM _ WHERE ((_ = _) AND (_ =...

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#24500]
The panic is coming from the second `panic()` in this code:

```
	// Only equivalency group representatives show up in weakKeys.
	for _, k := range pp.weakKeys {
		for c, ok := k.Next(0); ok; c, ok = k.Next(c + 1) {
			if repr := pp.eqGroups.Find(c); repr != c {
				panic(fmt.Sprintf("non-representative key set column %d (representative: %d)", c, repr))
			}
			if pp.constantCols.Contains(c) {
				panic(fmt.Sprintf("const column %d in key set %s", c, k))
			}
		}
	}
```

https://sentry.io/cockroach-labs/cockroachdb/issues/515498340/

```
*log.safeError: conn_executor.go:521: panic while executing 1 statements: SELECT _ FROM _ WHERE ((_ = _) AND (_ = $1)) AND (_ = _): caused by <redacted>
  File "github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go", line 489, in func2
  File "github.com/cockroachdb/cockroach/pkg/sql/physical_props.go", line 159, in check
  File "github.com/cockroachdb/cockroach/pkg/sql/physical_props.go", line 607, in computeMatchInternal
  File "github.com/cockroachdb/cockroach/pkg/sql/physical_props.go", line 652, in trim
  File "github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go", line 583, in simplifyOrderings
...
(12 additional frame(s) were not displayed)

conn_executor.go:521: panic while executing 1 statements: SELECT _ FROM _ WHERE ((_ = _) AND (_ = $1)) AND (_ = _): caused by <redacted>
```

**Comment by a-robinson** [comment]
Also https://sentry.io/cockroach-labs/cockroachdb/issues/515541600/

**Comment by jviksne** [comment]
Hi, if that would help in any way I'd like to comment that I just now experienced a crash with similar error (or is this my error auto-reported?) which I could narrow down to a query like this:

select TRUE from my_table where parent_id='70a86ba8-9dc0-482c-a959-b579a626b188' and child_id=parent_id;

The problem appears to be related to the following index:
CREATE UNIQUE INDEX ON my_table(delete_id, parent_id, child_id);

With no indexes present it works fine, when I add this index to the table then it crashes even with no data in the table. The child_id and parent_id are both UUID type columns. If I drop either of the conditions or replace child_id=parent_id with child_id='70a86ba8-9dc0-482c-a959-b579a626b188' (which can be used as a workaround) then it does not crash. It's reproducible with a new table.

Here is what I have in the console upon the crash:
*
* ERROR: [n1,client=127.0.0.1:57777,user=root] a SQL panic has occurred while executing "SELECT true FROM my_table WHERE (parent_id = '70a86ba8-9dc0-482c-a959-b579a626b123') AND (parent_id = child_id)": const column 1 in key set (1,6)
*
*
* ERROR: [n1,client=127.0.0.1:57777,user=root] a panic has occurred!
*
panic while executing 1 statements: SELECT _ FROM _ WHERE (_ = _) AND (_ = _); caused by const column 1 in key set (1,6)

goroutine 611 [running]:
runtime/debug.Stack(0x1ef2de0, 0xc047089b00, 0x3)
        /usr/local/go/src/runtime/debug/stack.go:24 +0xae
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x1ef2de0, 0xc047089b00, 0xc04256e000, 0x1af2960, 0xc0479d4cc0, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:195 +0xac
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc048823000, 0x1ef2de0, 0xc047089b00, 0x18be720, 0xc04777e4c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:521 +0x2dc
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc048823000, 0x1ef2de0, 0xc047089b00)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0x68
panic(0x18be720, 0xc04777e4c0)
        /usr/local/go/src/runtime/panic.go:505 +0x237
github.com/cockroachdb/cockroach/pkg/sql.physicalProps.check(0xc047970720, 0x3, 0x4, 0x2, 0x0, 0x6, 0x0, 0xc04777e490, 0x1, 0x1, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/physical_props.go:159 +0x8fd
github.com/cockroachdb/cockroach/pkg/sql.physicalProps.computeMatchInternal(0xc047970720, 0x3, 0x4, 0x2, 0x0, 0x6, 0x0, 0xc04777e490, 0x1, 0x1, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/physical_props.go:607 +0x67
github.com/cockroachdb/cockroach/pkg/sql.(*physicalProps).trim(0xc0476c38b0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/physical_props.go:652 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.(*planner).simplifyOrderings(0xc0488233e0, 0x1ee9520, 0xc0476c3800, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:687 +0x2434
github.com/cockroachdb/cockroach/pkg/sql.(*planner).simplifyOrderings(0xc0488233e0, 0x1ee9420, 0xc048000000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:690 +0x2121
github.com/cockroachdb/cockroach/pkg/sql.(*planner).expandPlan(0xc0488233e0, 0x1ef2ea0, 0xc0479d4810, 0x1ee9420, 0xc048000000, 0x1ee9420, 0xc048000000, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:38 +0x11e
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optimizePlan(0xc0488233e0, 0x1ef2ea0, 0xc0479d4810, 0x1ee9420, 0xc048000000, 0xc04808f4bc, 0x1, 0x1, 0x1ee9420, 0xc048000000, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/optimize.go:47 +0xf2
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc0488233e0, 0x1ef2ea0, 0xc0479d4810, 0x1ef56a0, 0xc049320100, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:323 +0x2f1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc048823000, 0x1ef2ea0, 0xc0479d4810, 0x1ef56a0, 0xc049320100, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:528 +0xb1b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc048823000, 0x1ef2ea0, 0xc0479d4810, 0x1ef56a0, 0xc049320100, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:342 +0x9ce
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc048823000, 0x1ef2ea0, 0xc0479d4810, 0x1ef56a0, 0xc049320100, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:83 +0x35f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc048823000, 0x1ef2de0, 0xc047089b00, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:911 +0x1d32
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0422ce000, 0x1ef2de0, 0xc047089b00, 0xc0482f204f, 0x7, 0xc0482f200d, 0x4, 0xc0482f2038, 0xd, 0x1edc260, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491 +0xfa5
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc0422ce000, 0x1ef2de0, 0xc047089b00, 0xc04230b880, 0x5400, 0x15000, 0xc042ab4190, 0xc046fd3e50, 0xc047dd97d0, 0xc046fd3e40)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0x103
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf1b

*
* ERROR: [n1,client=127.0.0.1:57777,user=root] Reported as error e3a2d399658d4e53b8f6cb14aedc867f
*

As a side note - in Windows after the crash I have to delete a temp directory (something like cockroach-temp780312239) under cockroach-data, otherwise the DB complains about a temp file locked by some other process and fails to start.

**Comment by RaduBerinde** [comment]
@jviksne thank you for the information! Can you share the schema of the table (SHOW CREATE TABLE). I'm in particular interested in what the primary key is.

**Comment by RaduBerinde** [comment]
Never mind, I can reproduce it.
```
root@127.166.243.53:34177/demo> create table cucu (a INT, b INT, c INT);
root@127.166.243.53:34177/demo> create unique index on cucu (a, b, c);
root@127.166.243.53:34177/demo> select true from cucu where b=1 and b=c;
```

**Comment by RaduBerinde** [comment]
@jviksne I will have a fix in 2.0.1.

In the meantime, as a workaround you can specify the value explicitly:
`where parent_id='70a86ba8-9dc0-482c-a959-b579a626b188' and child_id='70a86ba8-9dc0-482c-a959-b579a626b188'`

**Comment by jviksne** [comment]
Great, thanks! Yes, I figured that, and had this query just in a couple of places.
